File: api/models/analysis.py
"""Analysis status models for tracking async video analysis."""

from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel

AnalysisStatus = Literal["pending", "extracting", "analyzing", "annotating", "completed", "failed"]


class AnalysisState(BaseModel):
    """Full analysis state stored in status.json."""

    video_id: str
    status: AnalysisStatus
    progress: int  # 0-100
    error: Optional[str] = None
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    result_path: Optional[str] = None
    analysis_type: Optional[str] = None
    frames_dir: Optional[str] = None
    annotated_dir: Optional[str] = None


class AnalysisResponse(BaseModel):
    """Response model for analysis API endpoints."""

    video_id: str
    status: AnalysisStatus
    progress: int
    message: str

File: web/src/api/analysis.ts
import axios from 'axios'

const API_BASE = import.meta.env.VITE_API_BASE_URL || 'http://localhost:8000'

export interface AnalysisState {
  video_id: string
  status: 'pending' | 'extracting' | 'analyzing' | 'annotating' | 'completed' | 'failed'
  progress: number
  error?: string
  result_path?: string
}

export async function startAnalysis(videoId: string): Promise<{ video_id: string; status: AnalysisState['status'] }> {
  const response = await axios.post(`${API_BASE}/api/analyze/${videoId}`)
  return response.data
}

export async function getAnalysisStatus(videoId: string): Promise<AnalysisState> {
  const response = await axios.get(`${API_BASE}/api/status/${videoId}`)
  return response.data
}

export async function retryAnalysis(videoId: string): Promise<{ video_id: string; status: AnalysisState['status'] }> {
  const response = await axios.post(`${API_BASE}/api/retry/${videoId}`)
  return response.data
}